### PR TITLE
[error] Print JS function where error was thrown

### PR DESCRIPTION
**Pull request by jprestwo**
Signed-off-by: James Prestwood <james.prestwood@intel.com>

### CODE REVIEW

File: src/zjs_error.c [hunk 1]
```diff
@@ -80,6 +80,63 @@ void zjs_error_cleanup()
     }
 }
 
+static char* construct_message(jerry_value_t this,
+                                      jerry_value_t func,
+                                      const char *message)
```
Comment (grgustaf) [inline src/zjs_error.c]: Indent looks wrong, move func up to first line, tab message to the right place.

File: src/zjs_error.c [hunk 2]
```diff
@@ -80,6 +80,63 @@ void zjs_error_cleanup()
...
+{
+    jerry_size_t size = 64;
+    char name[size];
+    jerry_value_t keys_array = jerry_get_object_keys(this);
```
Comment (grgustaf) [inline src/zjs_error.c]: Should probably check for an error.

File: src/zjs_error.c [hunk 3]
```diff
@@ -80,6 +80,63 @@ void zjs_error_cleanup()
...
+    uint32_t arr_length = jerry_get_array_length(keys_array);
+    int i;
+    for (i = 0; i < arr_length; ++i) {
+        jerry_value_t val = jerry_get_property_by_index(keys_array, i);
+        zjs_copy_jstring(val, name, &size);
+        jerry_value_t func_val = zjs_get_property(this, name);
```
Comment (grgustaf) [inline src/zjs_error.c]: I think you can probably call jerry_get_property_by_index(this, i) to get func_val, and then you don't need to copy the string every time. If it matches, then copy the string and save the name.

File: src/zjs_error.c [hunk 4]
```diff
@@ -80,6 +80,63 @@ void zjs_error_cleanup()
...
+        if (func_val == func) {
+            break;
+        }
+        jerry_release_value(val);
+        jerry_release_value(func_val);
+    }
+    jerry_release_value(keys_array);
+
+    int nlen = strlen(name);
```
Comment (grgustaf) [inline src/zjs_error.c]: You already have size returned above so you should be able to skip this strlen.

File: src/zjs_error.c [hunk 5]
```diff
@@ -80,6 +80,63 @@ void zjs_error_cleanup()
...
+    char *msg = zjs_malloc(strlen(message) + nlen + 4);
+
+    memcpy(msg, name, nlen);
+    msg[nlen] = '(';
+    msg[nlen + 1] = ')';
+    msg[nlen + 2] = ':';
+    memcpy(msg + nlen + 3, message, strlen(message));
+    msg[strlen(message) + nlen + 3] = '\0';
```
Comment (grgustaf) [inline src/zjs_error.c]: Save the strlen(message) above rather than calling it 3x.

**Comment by grgustaf** [comment]
Otherwise, looks good. I guess we're not using it yet, but before we do we should make this macro do this only on a certain debug level. A lower level would leave the function names out and call the original function.

File: src/zjs_error.c (hunk 3 above)
Comment (jprestwo) [inline src/zjs_error.c]: I tried this and the function named ended up beings "X". Not sure why but it didn't seem to work.

File: src/zjs_error.c (hunk 4 above)
Comment (grgustaf) [inline src/zjs_error.c]: You can still make this change I think.

**Comment by grgustaf** [comment]
+1

**Pull request merged**